### PR TITLE
Filter out non-library returns in branches

### PR DESCRIPTION
**Pull request by casperboone**
This PR filters out non library return/throw statements that were retained by the value filter.

### CODE REVIEW

File: modules/pipeline/jimple-library-usage-graph-generator/src/test/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/IntegrationTest.kt [hunk 1]
```diff
@@ -149,17 +149,21 @@ internal class IntegrationTest : Spek({
...
-                            )
-                        )
+                        node<JAssignStmt>()
                     )
                 ),
                 libraryUsageGraph
             )
         }
+
+        it("filters out a class containing an if with method exitting return statements") {
```
Comment (gandreadis) [inline modules/pipeline/jimple-library-usage-graph-generator/src/test/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/IntegrationTest.kt]: I don't think there are no return statements (in Jimple) that aren't method-exitting - perhaps remove that part?

File: modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt [hunk 2]
```diff
@@ -55,48 +55,66 @@ private class BranchStatement(private val body: Body, val statement: Unit) {
...
 
-    private fun findBranchStatementEnd(): soot.Unit? {
-        // Return type is fully classified because of false positives by static analysis tools
+        private fun findBranchStatementEnd(): soot.Unit? {
+            // Return type is fully classified because of false positives by static analysis tools
 
-        val bodiesTillMethodEnd = cfg.getSuccsOf(statement).map { collectSuccessors(cfg, it) }
-        val intersectedBodies =
-            bodiesTillMethodEnd.fold(bodiesTillMethodEnd[0], { acc, list -> acc.intersect(list).toMutableList() })
+            val bodiesTillMethodEnd = cfg.getSuccsOf(statement).map { collectSuccessors(cfg, it) }
```
Comment (gandreadis) [inline modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: *Until

File: modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt [hunk 3]
```diff
@@ -55,48 +55,66 @@ private class BranchStatement(private val body: Body, val statement: Unit) {
...
-    }
+            return intersectedBodies[0]
+        }
 
-    private fun collectSuccessors(
-        cfg: UnitGraph, unit: Unit, collection: MutableList<Unit> = mutableListOf()
-    ): MutableList<Unit> {
-        if (collection.contains(unit)) return collection
+        private fun branchesContainOnlyReturnOrThrowStatements(branchTargets: List<Unit>) = branchTargets.all {
+            cfg.getSuccsOf(it).size == 0 && when (it) {
```
Comment (JSAbrahams) [inline modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: You can use `isEmpty` here

Comment (gandreadis) [reply modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: Fixed for @casperboone

File: modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt (hunk 2 above)
Comment (casperboone) [reply modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: Out of scope for this PR

File: modules/pipeline/jimple-library-usage-graph-generator/src/test/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/IntegrationTest.kt (hunk 1 above)
Comment (casperboone) [reply modules/pipeline/jimple-library-usage-graph-generator/src/test/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/IntegrationTest.kt]: You're correct, however, the fact that the return statements means that is exitted earlier than with a 'common end' statement after the branching statement (if) is very crucial. I think it helps the understanding of the reader of the test to see that it is about method exitting statements. The fact that these actually are return statements is actually slightly less relevant.

File: modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt (hunk 2 above)
Comment (FWDekker) [reply modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: [citation needed]

Comment (casperboone) [reply modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: "Out of scope for this PR"

Comment (JSAbrahams) [reply modules/pipeline/jimple-library-usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/pipeline/usagegraphgenerator/jimple/filters/BranchStatementFilter.kt]: *slow clap*

**Pull request merged**